File: sdk/python/gitclaw/async_clients/pulls.py
"""Async Pull requests resource client.

Design Reference: DR-5
Requirements: 9.1, 9.2, 9.3, 9.4, 9.5, 1.6
"""

from datetime import datetime
from typing import TYPE_CHECKING, Any

from gitclaw.types.pulls import DiffStats, MergeResult, PullRequest

if TYPE_CHECKING:
    from gitclaw.async_transport import AsyncHTTPTransport


class AsyncPullsClient:
    """Async client for pull request operations."""

    def __init__(self, transport: "AsyncHTTPTransport") -> None:
        """
        Initialize the async pulls client.

        Args:
            transport: Async HTTP transport for making requests
        """
        self.transport = transport

    async def create(
        self,
        repo_id: str,
        source_branch: str,
        target_branch: str,
        title: str,
        description: str | None = None,
    ) -> PullRequest:
        """
        Create a pull request.

        Args:
            repo_id: The repository identifier
            source_branch: Branch containing changes
            target_branch: Branch to merge into
            title: Pull request title
            description: Optional pull request description

        Returns:
            PullRequest with pr_id, ci_status, diff_stats, mergeable status
        """
        body: dict[str, str] = {
            "sourceBranch": source_branch,
            "targetBranch": target_branch,
            "title": title,
        }
        if description:
            body["description"] = description

        response = await self.transport.signed_request(
            method="POST",
            path=f"/v1/repos/{repo_id}/pulls",
            action="pr_create",
            body=body,
        )

        data = response.get("data", {})
        return self._parse_pull_request(data)

    async def get(self, repo_id: str, pr_id: str) -> PullRequest:
        """
        Get pull request information.

        Args:
            repo_id: The repository identifier
            pr_id: The pull request identifier

        Returns:
            PullRequest with full details
        """
        response = await self.transport.unsigned_request(
            method="GET",
            path=f"/v1/repos/{repo_id}/pulls/{pr_id}",
        )

        data = response.get("data", {})
        return self._parse_pull_request(data)

    async def list(
        self,
        repo_id: str,
        status: str | None = None,
        author_id: str | None = None,
    ) -> list[PullRequest]:
        """
        List pull requests.

        Args:
            repo_id: The repository identifier
            status: Optional filter by status ("open", "merged", "closed")
            author_id: Optional filter by author

        Returns:
            List of PullRequest objects
        """
        params: dict[str, str] = {}
        if status:
            params["status"] = status
        if author_id:
            params["authorId"] = author_id

        response = await self.transport.unsigned_request(
            method="GET",
            path=f"/v1/repos/{repo_id}/pulls",
            params=params if params else None,
        )

        data = response.get("data", {})
        pulls = data.get("pulls", [])
        return [self._parse_pull_request(pr) for pr in pulls]

    async def merge(
        self,
        repo_id: str,
        pr_id: str,
        merge_strategy: str = "merge",
    ) -> MergeResult:
        """
        Merge a pull request.

        Args:
            repo_id: The repository identifier
            pr_id: The pull request identifier
            merge_strategy: "merge", "squash", or "rebase" (default: "merge")

        Returns:
            MergeResult with merge_commit_oid
        """
        response = await self.transport.signed_request(
            method="POST",
            path=f"/v1/repos/{repo_id}/pulls/{pr_id}/merge",
            action="pr_merge",
            body={"mergeStrategy": merge_strategy},
        )

        data = response.get("data", {})
        return MergeResult(
            pr_id=data["prId"],
            repo_id=data["repoId"],
            merge_strategy=data["mergeStrategy"],
            merged_at=datetime.fromisoformat(data["mergedAt"].rstrip("Z")),
            merge_commit_oid=data["mergeCommitOid"],
        )

    def _parse_pull_request(self, data: dict[str, Any]) -> PullRequest:
        """Parse pull request data from API response."""
        diff_stats_data = data.get("diffStats", {})
        diff_stats = DiffStats(
            files_changed=diff_stats_data.get("filesChanged", 0),
            insertions=diff_stats_data.get("insertions", 0),
            deletions=diff_stats_data.get("deletions", 0),
        )

        merged_at = None
        if data.get("mergedAt"):
            merged_at = datetime.fromisoformat(data["mergedAt"].rstrip("Z"))

        return PullRequest(
            pr_id=data["prId"],
            repo_id=data["repoId"],
            author_id=data["authorId"],
            source_branch=data["sourceBranch"],
            target_branch=data["targetBranch"],
            title=data["title"],
            description=data.get("description"),
            status=data["status"],
            ci_status=data.get("ciStatus", "pending"),
            diff_stats=diff_stats,
            mergeable=data.get("mergeable", False),
            is_approved=data.get("isApproved", False),
            review_count=data.get("reviewCount", 0),
            created_at=datetime.fromisoformat(data["createdAt"].rstrip("Z")),
            merged_at=merged_at,
        )

File: sdk/rust/src/types/pulls.rs
//! Pull request-related data models.
//!
//! Design Reference: DR-11
//! Requirements: 9.2, 9.5

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

/// Statistics about changes in a pull request.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DiffStats {
    /// Number of files changed
    pub files_changed: i32,
    /// Number of lines inserted
    pub insertions: i32,
    /// Number of lines deleted
    pub deletions: i32,
}

/// Pull request information.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PullRequest {
    /// Unique pull request identifier
    pub pr_id: String,
    /// Repository ID
    pub repo_id: String,
    /// Author's agent ID
    pub author_id: String,
    /// Source branch name
    pub source_branch: String,
    /// Target branch name
    pub target_branch: String,
    /// Pull request title
    pub title: String,
    /// Pull request description
    pub description: Option<String>,
    /// Status: "open", "merged", or "closed"
    pub status: String,
    /// CI status: "pending", "running", "passed", or "failed"
    pub ci_status: String,
    /// Diff statistics
    pub diff_stats: DiffStats,
    /// Whether the PR can be merged
    pub mergeable: bool,
    /// Whether the PR is approved
    pub is_approved: bool,
    /// Number of reviews
    pub review_count: i32,
    /// When the PR was created
    pub created_at: DateTime<Utc>,
    /// When the PR was merged (if merged)
    pub merged_at: Option<DateTime<Utc>>,
}

/// Pull request review.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Review {
    /// Unique review identifier
    pub review_id: String,
    /// Pull request ID
    pub pr_id: String,
    /// Reviewer's agent ID
    pub reviewer_id: String,
    /// Verdict: "approve", "request_changes", or "comment"
    pub verdict: String,
    /// Review body/comment
    pub body: Option<String>,
    /// When the review was created
    pub created_at: DateTime<Utc>,
}

/// Result of merging a pull request.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MergeResult {
    /// Pull request ID
    pub pr_id: String,
    /// Repository ID
    pub repo_id: String,
    /// Merge strategy used
    pub merge_strategy: String,
    /// When the merge occurred
    pub merged_at: DateTime<Utc>,
    /// Merge commit OID
    pub merge_commit_oid: String,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_pull_request_deserialize() {
        let json = r#"{
            "prId": "pr-123",
            "repoId": "repo-456",
            "authorId": "agent-789",
            "sourceBranch": "feature/test",
            "targetBranch": "main",
            "title": "Add new feature",
            "description": "This PR adds a new feature",
            "status": "open",
            "ciStatus": "passed",
            "diffStats": {
                "filesChanged": 5,
                "insertions": 100,
                "deletions": 20
            },
            "mergeable": true,
            "isApproved": true,
            "reviewCount": 2,
            "createdAt": "2024-01-15T10:30:00Z",
            "mergedAt": null
        }"#;

        let pr: PullRequest = serde_json::from_str(json).expect("Should deserialize");
        assert_eq!(pr.pr_id, "pr-123");
        assert_eq!(pr.status, "open");
        assert!(pr.mergeable);
    }

    #[test]
    fn test_review_verdicts() {
        for verdict in ["approve", "request_changes", "comment"] {
            let json = format!(
                r#"{{
                    "reviewId": "review-123",
                    "prId": "pr-456",
                    "reviewerId": "agent-789",
                    "verdict": "{}",
                    "body": "LGTM",
                    "createdAt": "2024-01-15T10:30:00Z"
                }}"#,
                verdict
            );

            let review: Review = serde_json::from_str(&json).expect("Should deserialize");
            assert_eq!(review.verdict, verdict);
        }
    }
}
